datafusion/src/physical_plan/metrics.rs

```diff
@@ -0,0 +1,461 @@
+    /// across partitions or None if no metric is present
+    pub fn output_rows(&self) -> Option<usize> {
+        self.sum(|metric| matches!(metric.kind(), MetricKind::OutputRows))
+    }
+
+    /// convenience: return the amount of CPU time spent, aggregated
+    /// across partitions or None if no metric is present
+    pub fn cpu_time(&self) -> Option<usize> {
+        self.sum(|metric| matches!(metric.kind(), MetricKind::CPUTime))
+    }
+
+    /// Sums the values for metrics for which `f(metric)` returns
+    /// true, and returns the value. Returns None if no metrics match
+    /// the predicate.
+    pub fn sum<F>(&self, mut f: F) -> Option<usize>
+    where
+        F: FnMut(&SQLMetric) -> bool,
+    {
+        let mut iter = self
+            .metrics
+            .iter()
+            .filter(|metric| f(metric.as_ref()))
+            .peekable();
+
+        if iter.peek().is_none() {
+            None
+        } else {
+            Some(iter.map(|metric| metric.value()).sum())
+        }
+    }
+
+    /// Returns returns a new derived `MetricsSet` where all metrics
+    /// that had partition=`Some(..)` have been aggregated
+    /// together. The resulting `MetricsSet` has all metrics with `Partition=None`
+    pub fn aggregate_by_partition(&self) -> Self {
+        let mut map = HashMap::new();
+
+        // There are all sorts of ways to make this more efficient
+        for metric in &self.metrics {
+            let key = (metric.kind.clone(), metric.labels.clone());
+            map.entry(key)
+                .and_modify(|accum: &mut SQLMetric| {
+                    accum.set(accum.value() + metric.value())
+                })
+                .or_insert_with(|| {
+                    // accumuate with no partition
+                    let partition = None;
+                    let accum = SQLMetric::new_with_labels(
+                        metric.kind().clone(),
+                        partition,
+                        metric.labels().to_vec(),
+                    );
+                    accum.set(metric.value());
+                    accum
+                });
+        }
+
+        let new_metrics = map
+            .into_iter()
+            .map(|(_k, v)| Arc::new(v))
+            .collect::<Vec<_>>();
+
+        Self {
+            metrics: new_metrics,
+        }
+    }
+}
+
+impl Display for MetricsSet {
+    /// format the MetricsSet as a single string
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut is_first = true;
+        for i in self.metrics.iter() {
+            if !is_first {
+                write!(f, ", ")?;
+            } else {
+                is_first = false;
+            }
+
+            write!(f, "{}", i)?;
+        }
+        Ok(())
+    }
+}
+
+/// A set of SQLMetrics that can be added to as partitions
+/// execute. Designed to be a convenience for operator implementation
+#[derive(Default, Debug)]
+pub struct SharedMetricsSet {
```

Comment:
The name of this confused me for a while. At least on my first pass Shared made me think shared between operators, when this clearly isn't the case. I don't really have a better suggestion for a name, but thought I'd mention it. Maybe OperatorMetricsSet or something, I don't know...

Reply:
Renamed to `ExecutionPlanMetricsSet` and added some clarifying doc comments